Addressed AltLoc CIF bug in structure_from_doc

danny305 opened the pull request:
Also refactored a little bit of your cif.cc file to use unique_ptrs.

mittinatten commented on .gitmodules:
```diff
@@ -1,3 +1,3 @@
 [submodule "third-party/gemmi"]
 	path = third-party/gemmi
-	url = git@github.com:project-gemmi/gemmi.git
+	url = https://github.com/project-gemmi/gemmi.git
```
Yes, this is probably better

danny305 replied:
It would not let me run git submodule update --init without this fix.

mittinatten commented on src/config.h:
```diff
@@ -0,0 +1,244 @@
+/* config.h.  Generated from config.h.in by configure.  */
```
This file depends on the user environment, so it should be generated by the build system, and not be included in the library.

danny305 replied:
Yeah I know. This was just a hack I did to get it working on my system. 

Generating this file from the user environment is something I was planning on implementing with cmake later. 

I was just prioritizing getting it building on my system so I could focus on cif stuff.

mittinatten commented on src/cif.cc:
```diff
@@ -85,16 +86,13 @@ structure_from_doc(const gemmi::cif::Document &doc,
-            if ((currentAltId != '?' && prevAltId == '?') || (currentAltId == '?')) {
-                prevAltId = currentAltId;
-            } else if (currentAltId != '?' && currentAltId != prevAltId) {
+            if (currentAltId != '.' && currentAltId != 'A') {
```
Are we sure alt ids always begin with 'A'? (If so, that is a nice simplification that should be used in the pdb code as well)

danny305 replied:
So Alt ID is either a `.` if there is only one rotamer/conformer for that residue. 

If there are several they are distinguished by A, B, C, etc. And in our case we default to the first one, which is A.

I hope this is enough. This is where that test set of pdb/cif files comes in handy.

mittinatten replied:
Right, the other solution just takes the first one, even if A is missing for some reason. But I guess we'll find out in our tests as you say. This is cleaner if it works.

mittinatten commented on src/cif.cc:
```diff
@@ -1,13 +1,14 @@
 #include "cif.hh"
 
-static std::set<int> *
+static std::unique_ptr<std::set<int>> 
```
This is great, C++ is so much nicer now than last time I used it :)

mittinatten reviewed (changes requested):
So I found out why you had problems building this with autotools: `make clean` didn't remove the `.deps` files. I have fixed it in the target branch. Replacing autotools is such a major change that I would prefer to do it separately, and it requires thorough testing (and I'm not sure it's worth it, since the current setup works quite well, although somewhat arcane).

The changes to the actual app logic look nice (see my one question to cif.cc though). Could you edit the PR to only include the changes to cif.cc and the .submodules (and preferably rebase it down to 1 or 2 commits)?

danny305 commented:
I have rebased my pr branch to remove all the cmake stuff.

I pulled in the changes you made for autotools to work on my computer. I updated the c++ minimum standard to c++14 so it would successfully compile.

mittinatten commented:
The git -> https submodule change wasn't here, but it turns out is needed in travis for CI as well. So I added it myself. (not sure if this link is publicly available: https://travis-ci.org/github/mittinatten/freesasa/jobs/759850250)

danny305 commented:
Good catch!
